File: Haskell/pl20140703_demux.hs
-- Consider a variant of the problem seen in Exercise 1: an unsorted list can contain list of elements of some
-- type, integer numbers, or the special value End.
-- E.g. (in a pseudo-Haskell syntax) [3, [6, 6, 1], 4, [1, 2], -2, End, 9].

-- Define the data structure for the unsorted list, and Demuxed, analogous to the structure introduced in Exercise
-- 1 (i.e. with two fields, one containing an integer value, the sum of the integer elements found, and a list
-- of all the found lists).

data UList a = L [a] | V a | End

data Demuxed a = Demuxed Int [[a]]

-- Define the demux function, that takes an unsorted list l and builds up a demuxed data structure containing
-- the processed data of l (only that before End, if present). demux must be strict (i.e. non lazy).

demux l = demux' l 0 []
		 	where demux' ((V l):ls) n v = let n' = l + n in seq n' demux' ls n' v
		 		  demux' ((L l):ls) n v = let v' = (v ++ l) in seq demux' ls n v'
		 		  demux' (End:ls) n v = Demuxed n v
		 		  demux' [] n v = Demuxed n v

File: Prolog/pl20140703_demux.pl
% Consider a variant of the problem seen in Exercise 1: an unsorted list can contain either atoms or integer
% numbers. E.g. [1,3,house,4,of,-7,cards,end,-12].
% Define a demux predicate, that has an unsorted list as first argument, the sum n of the numbers present
% in the unsorted list as second argument, and puts all the symbols in the list v as third argument. demux must
% stop if an atom end is found, and must be optimized with cut.
% E.g. for the previous example list n = 1 and v = [house,of,cards].

demux([end|_],0,[]) :- !.
demux([X|Xs],N,V) :- number(X), !, demux(Xs,N1,V), !, N is N1 + X.
demux([X|Xs],N,[X|V]) :- !, demux(Xs,N,V).
demux([],0,[]) :- !.

% Define a shuffle predicate, that takes three lists n, a, m, where n is a list of number, a is a list of atoms,
% and m is a list of numbers and atoms containing the elements of n and a, by maintaining their relative order
% (e.g. shuffle([1,2],[house,next,foot],[house,1,next,foot,2]) must hold). shuffle must be able to
% shuffle two lists together, e.g. to obtain as output the third list in the example.

shuffle([],[],[]).
shuffle(V,[X|Ys],[X|Xs]) :- atom(X), shuffle(V,Ys,Xs).
shuffle([X|Ys],V,[X|Xs]) :- number(X), shuffle(Ys,V,Xs).
